### ISSUE TITLE
[RFC] Degree Padding Scheduler

### ISSUE THREAD

**Issue by BarclayII** [dmlc/dgl#777]
## 🚀 Feature
Provide an option to schedule reducers with degree padding strategy, i.e. padding the incoming messages to equal length.

Optionally, provide a hybrid option of both degree bucketing and degree padding, i.e. divide the nodes into coarser buckets, each of which corresponds to a range of degrees, and perform degree padding in the coarser buckets.

## Motivation

Current scheduling strategy for custom reduce UDFs is degree bucketing, where the nodes with the same number of incoming messages are stacked together and passed into the reduce UDF at a time.  This would often make running sequential reducer UDFs (e.g. LSTMs) slow, especially if all the in-degree values are lower than a small value.  For instance, running a LSTM 10 times would be certainly slower than running a LSTM once on messages padded to length 10.

The inefficiency of running LSTM-based GraphSAGE is observed by @yzh119 .

In real world graphs where power-law is often observed, the maximum in-degree could be very big, while the distribution of in-degrees would be long-tailed.  In this case, a hybrid strategy is often preferred (e.g. merge all nodes with in-degree less than 10 into one bucket, and bucket all nodes with in-degrees greater than 10 separately).

For now, we can implement a manual version of degree padding (and hybrid) reducer scheduler where the buckets would be manually specified by the developers.

## Pitch

We use decorator to specify the strategy of a reducer UDF (or any callable):

```python
# Specifying strategy for a function
@dgl.reduce_scheduler(strategy=strategy_name, **kwargs)
def reduce_func(nodes):
    # ...

# Or specifying for a class
@dgl.reduce_scheduler(strategy=strategy_name, **kwargs)
class Reduce(nn.Module):
    # ...
```

Such decorator would introduce a new member **`mailbox_length`** to the `nodes` argument, which is an Int64 vector with the same number of rows as `nodes.mailbox[...]`.

For instance, with an LSTM reducer (code in PyTorch):

```python
@dgl.reduce_scheduler(strategy='padding', bucket_split=None)
class LSTMAgg(nn.Module):
    def forward(self, nodes):
        # padded, (batch_size, max_messages, dim)
        m = nodes.mailbox['m']
        # (batch_size,) LongTensor.
        # for each row i, the mailbox entries
        #
        #     nodes.mailbox['m'][i, :nodes.mailbox_length['m']]
        #
        # are meaningful.  Other entries are padded with zeros.
        length = nodes.mailbox_length['m']

        pack_seq = pack_padded_sequence(
            m, length, batch_first=True, enforce_sorted=False)
        output = self.lstm(pack_seq)
        pad_out = pad_packed_sequence(output, batch_first=True)
        return output.sum(1) / length[:, None].float()

reduce_func = LSTMAgg()

g.update_all(
        message_func=fn.copy('h', 'm'),
        reduce_func=reduce_func)
```

Right now, we could support two strategies:

* `dgl.reduce_scheduler(strategy='padding', bucket_split=[10, 20])`:
  The scheduler would bucket in-degree from 1 to 10 (exclusive) into one bucket, 10 to 20 (exclusive) into another bucket, and all in-degrees bigger than 20 into separate buckets.  The nodes in the same bucket are then padded to the maximum number of incoming messages.
  If `bucket_split` is `None` (default), then all in-degrees are merged into a single bucket.
* If no decorator is given, then the reducer UDF still defaults to degree bucketing.
* In (far?) future, the buckets could be automatically determined by the system.

## Additional context

This is in response to the issue #736 .

EDIT: `edges` -> `nodes`

**Comment by VoVAllen** [comment]
Seems a good feature to me. Some concerns:
- The predefined bucketing strategy should be provided with this API instead of doing this in the future. It's too detailed for user to do this. As well as some speed profiling results.
- `edges.mask` would be a useful API, which returns the corresponding mask 
- Do we have enough time to do this now?

**Comment by BarclayII** [comment]
> * The predefined bucketing strategy should be provided with this API instead of doing this in the future. It's too detailed for user to do this. As well as some speed profiling results.

The problem is that reasonable bucketing strategies differ within each dataset and each graph, because the in-degree distributions of different graphs vary.  I think deciding an optimal strategy of bucketing and padding is a research problem.

For now, we could have some heuristic strategies, like "put 95% of the nodes ordered by in-degree into the same bucket", but I don't think that users could count on these heuristics.

> * `edges.mask` would be a useful API, which returns the corresponding mask

Sure.  `nodes.mask` would return a `batch_size x max_messages` boolean tensor in this case. 

> * Do we have enough time to do this now?

I could try.  Since this only involves scheduling with custom reducer UDFs, I don't think it will significantly change a lot of code.

**Comment by VoVAllen** [comment]
No need to be optimal, but to ensure the strategy is runnable (won't OOM) would be fine.
